### PR TITLE
e2e Tests

### PR DESCRIPTION
**Pull request by priyamsahoo**
The PR contains
- the test setup for end-to-end (e2e) tesing of the extension.
- basic testing for diagnostics using ansible-lint and syntax-check.

TODO:
- [x] Structure the test fixtures in a more appropriate way.
- [x] Generalize diagnostic test script.
- [x] Modify assert statements to use non-deprecated functions.

### CODE REVIEW

**Comment by ssbarnea** [comment]
I do not understand why you did not just add the next tests to the already existing `extest` suite. That is what runs when you execute `npm run ui-test` and already takes care about installing an isolated version of vscode and out extension. Basically you only need to add extra operations  to https://github.com/ansible/vscode-ansible/tree/main/src/ui-test

If I am correct, you should not need to modify anything inside `package.json`, no extra scripts.

File: src/test/testDiagnotics.test.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+describe('TEST FOR DIAGNOSTICS IN LOCAL ENVIRONMENT (without ee)', () => {
+  const docUri1 = getDocUri('diagnostics/1.yml');
+  const docUri2 = getDocUri('diagnostics/2.yml');
+
+  beforeEach(async () => {
+    await vscode.commands.executeCommand('workbench.action.closeAllEditors');
+  });
+
+  describe('Diagnostic test with ansible-lint', () => {
+    it('Test 1', async () => {
```
Comment (webknjaz) [inline src/test/testDiagnotics.test.ts:19]: The test names should contain a meaningful continuation of `it` and probably use the word `should`. So that it'd read `it should do/result in something`.
```suggestion
    it('should result in a result', async () => {
```

File: src/test/helper.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+export async function testDiagnostics(
+  docUri: vscode.Uri,
+  expectedDiagnostics: vscode.Diagnostic[]
+): Promise<void> {
+  const actualDiagnostics = vscode.languages.getDiagnostics(docUri);
+
+  assert.strictEqual(actualDiagnostics.length, expectedDiagnostics.length);
+
+  if (actualDiagnostics.length !== 0 && expectedDiagnostics.length !== 0) {
```
[the remark this answers is not in the document]
Comment (webknjaz) [reply src/test/helper.ts:63]: Even better. Since the check above already verifies that the lengths are equal, just check one.
```suggestion
  if (!actualDiagnostics.length) {
    return;
  }
```

File: src/test/helper.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    expectedDiagnostics.forEach((expectedDiagnostic, i) => {
```
Comment (webknjaz) [inline src/test/helper.ts:64]: Don't use `Array.prototype.forEach()` here. Just use the normal `for (let i in expectedDiagnostic) {}` loop.

File: src/test/testDiagnotics.test.ts (hunk 1 above)
Comment (ssbarnea) [reply src/test/testDiagnotics.test.ts:19]: I do agree that "test 1" does not say anything meaningful but neither `should result in a result` ;)

Comment (priyamsahoo) [reply src/test/testDiagnotics.test.ts:19]: I agree. I did not intend to have this message. It was only for testing purpose and then I forgot to remove from before pushing the PR.

File: src/test/helper.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (priyamsahoo) [reply src/test/helper.ts:63]: Thanks for the suggestion !

File: src/test/helper.ts (hunk 3 above)
Comment (priyamsahoo) [reply src/test/helper.ts:64]: Is there any specific reason for that ? 
Generally, I prefer writing this way when I do not want to break out of the loop.

File: src/test/testDiagnotics.test.ts (hunk 1 above)
Comment (priyamsahoo) [reply src/test/testDiagnotics.test.ts]: Done !

Comment (webknjaz) [reply src/test/testDiagnotics.test.ts]: > I do agree that "test 1" does not say anything meaningful but neither `should result in a result` ;)

Totally, it was mostly to demonstrate how hard it is to guess what the test expectation is if the author does not provide it.

File: src/test/helper.ts (hunk 3 above)
Comment (webknjaz) [reply src/test/helper.ts:88]: It just seems cleaner.

File: src/test/test.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+'use strict';
```
Comment (webknjaz) [inline src/test/test.js:1]: Do we actually need this JS counterpart of the TS file?

File: src/test/test.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+export const defaultAnsibleConfigurations = {
```
Comment (webknjaz) [inline src/test/test.ts:1]: Why is this duplicate declaration needed?

File: src/test/test.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  'executionEnvironment.image':
+    'quay.io/ansible/ansible-devtools-demo-ee:v0.1.0',
+
+  'executionEnvironment.pullPolicy': 'missing',
+  'python.activationScript': '',
+  'python.interpreterPath': 'python3',
+  'ansible.path': 'ansible',
+};
+
+// defaultAnsibleConfigurations.forEach((config) => {
```
Comment (webknjaz) [inline src/test/test.ts:19]: Maybe don't commit this?

File: src/test/testRunner.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+} from 'vscode-test';
+
+async function main(): Promise<void> {
+  try {
+    const executable = await downloadAndUnzipVSCode();
+    const cliPath = resolveCliPathFromVSCodeExecutablePath(executable);
+
+    // Install the latest released redhat.ansible extension
+    const installLog = cp.execSync(
+      `"${cliPath}" --install-extension redhat.ansible --force`
```
Comment (webknjaz) [inline src/test/testRunner.ts:16]: Why is this necessary?

File: src/test/testRunner.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    );
+    console.log(installLog.toString());
+
+    // Install the dependant extensions
```
Comment (webknjaz) [inline src/test/testRunner.ts:20]: Why can't this just come from the deps?

File: src/test/testScripts/diagnostics/testDiagnoticsAnsibleLocal.test.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    describe('Diagnostic test with ansible-lint', () => {
+      beforeEach(async () => {
+        await updateSettings('ansibleLint.enabled', true);
+      });
+
+      it('should complain about no task names', async () => {
+        await activate(docUri1);
+        await vscode.commands.executeCommand('workbench.action.files.save');
+
+        await sleep(2000); // Wait for the diagnostics to compute on this file
```
Comment (webknjaz) [inline src/test/testScripts/diagnostics/testDiagnoticsAnsibleLocal.test.ts:34]: Is there a more reliable way to wait?

File: src/test/testScripts/diagnostics/testDiagnoticsAnsibleLocal.test.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+        await testDiagnostics(docUri1, []);
+      });
+
+      // eslint-disable-next-line quotes
+      it("should complain about missing 'hosts' key", async function () {
```
Comment (webknjaz) [inline src/test/testScripts/diagnostics/testDiagnoticsAnsibleLocal.test.ts:84]: how about
```suggestion
      it('should complain about missing `hosts` key', async function () {
```

File: src/test/testScripts/hover/testHoverLocal.test.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+  resetDefaultSettings,
+} from '../../helper';
+
+export function testHoverLocal(): void {
+  describe('TEST FOR HOVER', () => {
+    const docUri1 = getDocUri('hover/1.yml');
+
+    beforeEach(async () => {
+      await resetDefaultSettings();
+      // await vscode.commands.executeCommand('workbench.action.closeAllEditors');
```
Comment (webknjaz) [inline src/test/testScripts/hover/testHoverLocal.test.ts:17]: Any comment on this?

File: src/test/testScripts/hover/testHoverLocal.test.ts [hunk 12]
```diff
@@ -0,0 +1,143 @@
+/* eslint-disable quotes */
```
Comment (webknjaz) [inline src/test/testScripts/hover/testHoverLocal.test.ts:1]: Any linting exception must have reasoning behind it.

**Review by webknjaz — CHANGES_REQUESTED** [review]
I've added some notes + add this to the CI plz.

File: .gitignore [hunk 13]
```diff
@@ -5,3 +5,5 @@ out
 .python-version
 test-resources
 test-resources-oldest
+.vscode-test
+e2eTestReport/
```
[the remark this answers is not in the document]
Comment (priyamsahoo) [reply .gitignore:9]: .vscode-test has a different purpose all together. It contains the downloaded vscode package for test. Are you suggesting to put e2eTestReport inside .vscode ? If that's the case, I don't think that's a great idea.

File: package.json [hunk 14]
```diff
@@ -448,7 +450,9 @@
     "webpack": "npm run clean && npm run compile && webpack --mode production --config ./webpack.config.js",
     "webpack:dev": "npm run clean && webpack --mode development --config ./webpack.config.js",
     "webpack:watch": "webpack --mode development --config ./webpack.config.js --watch",
-    "clean": "rimraf out/client out/server out/tsconfig.tsbuildinfo"
+    "clean": "rimraf out/client out/server out/tsconfig.tsbuildinfo",
+    "test-compile": "npm run clean && tsc -p ./",
+    "test:e2e": "npm run test-compile && node ./out/client/test/testRunner"
```
[the remark this answers is not in the document]
Comment (priyamsahoo) [reply package.json:455]: Thanks for the suggestion.

File: src/test/helper.ts [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * Activates the redhat.ansible extension
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export async function activate(docUri: vscode.Uri): Promise<any> {
+  const extension = vscode.extensions.getExtension('redhat.ansible');
+  const activation = await extension?.activate();
+
+  try {
```
[the remark this answers is not in the document]
Comment (priyamsahoo) [reply src/test/helper.ts:32]: Basically, there are 3 statements inside that and all of them are kind of linked to each other and serve the same purpose (open the doc and set language to 'ansible'). That's the reason I have kept it that way. I don't think I can further refactor it.

File: src/test/helper.ts [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+    doc = await vscode.workspace.openTextDocument(docUri);
+    editor = await vscode.window.showTextDocument(doc, {
+      preview: true,
+      preserveFocus: false,
+    });
+    await vscode.languages.setTextDocumentLanguage(doc, 'ansible');
+
+    await sleep(5000); // Wait for server activation
```
[the remark this answers is not in the document]
Comment (priyamsahoo) [reply src/test/helper.ts:40]: As of now, I don't have a better solution for this, since there's no such actionable indication when the server is initiated. Maybe, this can be discussed and a better implementation (if any), can be added in subsequent PRs.

File: src/test/helper.ts [hunk 17]
```diff
@@ -0,0 +1,120 @@
+import * as vscode from 'vscode';
+import * as path from 'path';
+import { assert } from 'chai';
+
+export let doc: vscode.TextDocument;
+export let editor: vscode.TextEditor;
+
+// Default ansible configurations
+export const defaultAnsibleConfigurations = {
```
[the remark this answers is not in the document]
Comment (priyamsahoo) [reply src/test/helper.ts:9]: Working on it. I want that to be the part of a separate PR.

File: src/test/helper.ts [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+  const ansibleConfiguration = vscode.workspace.getConfiguration('ansible');
+  return ansibleConfiguration.update(setting, value, false);
+}
+
+export async function resetDefaultSettings(): Promise<void> {
+  const ansibleConfiguration = vscode.workspace.getConfiguration('ansible');
+  Object.entries(defaultAnsibleConfigurations).forEach((config) => {
+    ansibleConfiguration.update(config[0], config[1], false);
+  });
+  await sleep(1000);
```
[the remark this answers is not in the document]
Comment (priyamsahoo) [reply src/test/helper.ts:76]: When the settings all reset, if we do not wait, they are not reflected in the config. This is one is tested. Resetting of configs, take some time. That's why, there's a wait statement over there.

File: src/test/test.js (hunk 4 above)
Comment (priyamsahoo) [reply src/test/test.js:1]: That's a mistake, this file should not be a part of the PR altogether, will remove it.

File: src/test/test.ts (hunk 5 above)
Comment (priyamsahoo) [reply src/test/test.ts:1]: Same here, `test.ts` and `test.js` are not actually the part of the PR. I will remove it. Thanks.

File: src/test/test.ts (hunk 6 above)
Comment (priyamsahoo) [reply src/test/test.ts:19]: Same as above. The files are unnecessary. I will remove them.

File: src/test/testRunner.ts (hunk 7 above)
Comment (priyamsahoo) [reply src/test/testRunner.ts:16]: When test starts for the first time and vscode is downloaded in the space, installations of the extensions need to be performed so that they can be tested. Plus `--force` flag is necessary so that the latest version of the extension is downloaded and all testings can be performed on it.

File: src/test/testRunner.ts (hunk 8 above)
Comment (priyamsahoo) [reply src/test/testRunner.ts:20]: Sorry, unable to understand it. Could you please elaborate?

File: src/test/testScripts/diagnostics/testDiagnoticsAnsibleLocal.test.ts (hunk 9 above)
Comment (priyamsahoo) [reply src/test/testScripts/diagnostics/testDiagnoticsAnsibleLocal.test.ts:34]: As of now, I could not find a better way to do it. Maybe if we find it in future, the changes can be made in the subsequent PRs.

File: src/test/testScripts/hover/testHoverLocal.test.ts (hunk 11 above)
Comment (priyamsahoo) [reply src/test/testScripts/hover/testHoverLocal.test.ts:17]: I will remove the commented code. Thanks.

File: src/test/testScripts/diagnostics/testDiagnoticsAnsibleLocal.test.ts (hunk 10 above)
Comment (priyamsahoo) [reply src/test/testScripts/diagnostics/testDiagnoticsAnsibleLocal.test.ts:84]: Thanks for the suggestion, this seems a better representation. Will do it this way.

File: src/test/testScripts/hover/testHoverLocal.test.ts (hunk 12 above)
Comment (priyamsahoo) [reply src/test/testScripts/hover/testHoverLocal.test.ts:1]: Yes. There's reason behind this as well.
But seems like if implemented the way you suggested in the previous comment, that linting exception can be removed from at least this file. Thanks for the suggestion.

File: .gitignore (hunk 13 above)
[the remark this answers is not in the document]
Comment (ssbarnea) [reply .gitignore:9]: I think we should start sticking temp folders inside `out/`. If I remember well we already have this folder and we can other things to it and avoid cluttering repository root with other folders.

I do remember even moving the test-resources folder there, maybe i forgot to remove them from gitignore or failed to push that change. The reason the same: avoiding temp folders in root level.

I am not sure if we can do the some about the `.vscode-test` folder but at least for the report we can clearly do it.

**Comment by ssbarnea** [comment]
@priyamsahoo @webknjaz Please mark threads as resolved if they were resolved, it would make it easier to realise which ares still need work or not.

**Review by ganeshrn — APPROVED** [review]

File: .gitignore (hunk 13 above)
[the remark this answers is not in the document]
Comment (priyamsahoo) [reply .gitignore:9]: Thanks for the suggestion. The `e2eTestReport` folder has been moved inside the `out` folder.

**Comment by ssbarnea** [comment]
I tried to run `npm run test:e2e` on macos and it failed to run with multiple tests. IMHO, review process should not even start before we have a running and passing job on GHA. I pushed a change for enabling it, so we should see results soon.

File: .github/workflows/ci.yaml [hunk 19]
```diff
@@ -29,31 +29,32 @@ jobs:
       fail-fast: false
       matrix:
         os:
-        - ubuntu-latest
-        - macos-11
-        - windows-latest
+          - ubuntu-latest
```
Comment (webknjaz) [inline .github/workflows/ci.yaml:32]: Plz don't add unnecessary indents to YAML sequences.

File: .github/workflows/ci.yaml [hunk 20]
```diff
@@ -29,31 +29,32 @@ jobs:
...
+          - windows-latest
         npm-target:
-        - test:ui
-        - test:ui-oldest
+          # - test:ui
+          # - test:ui-oldest
+          - test:e2e
         upload-artifact:
-        - false
+          - false
```
Comment (webknjaz) [inline .github/workflows/ci.yaml:40]: Looks like most of the changes to this file must be reverted.

**Review by ssbarnea — CHANGES_REQUESTED** [review]
Once we have the workaround for https://github.com/ansible/vscode-ansible/issues/373 added to it it should be ok.

**Comment by priyamsahoo** [comment]
All the smaller parts of this PR are merged successfully and the tests are passing in the CI as well.
- PR for hover e2e-test: https://github.com/ansible/vscode-ansible/pull/370
- PR for ansible diagnostic e2e-test: https://github.com/ansible/vscode-ansible/pull/382
- PR for YAML diagnostic e2e-test: https://github.com/ansible/vscode-ansible/pull/396
